mesonbuild/mesonlib.py

```diff
@@ -982,49 +1036,57 @@ def replace_if_different(dst, dst_tmp):
     else:
         os.unlink(dst_tmp)
 
-def listify(item: T.Any,
-            flatten: bool = True,
-            unholder: bool = False) -> T.List[T.Any]:
+
+@T.overload
+def unholder(item: 'ObjectHolder[_T]') -> _T: ...
+
+@T.overload
+def unholder(item: T.List['ObjectHolder[_T]']) -> T.List[_T]: ...
+
+@T.overload
+def unholder(item: T.List[_T]) -> T.List[_T]: ...
+
+@T.overload
+def unholder(item: T.List[T.Union[_T, 'ObjectHolder[_T]']]) -> T.List[_T]: ...
+
+def unholder(item):
```

Comment:
If you want the *implementation* to be type-checked, you need to annotate it as well (with a superset of all the individual overloads). Might be too annoying though.

Reply:
The docs say not to do this, and when I tried mypy complains about any signature I give it, complaining that one or more overloads doesn't match that signature. I'm going to leave this plain.

Reply:
> The docs say not to do this

The mypy docs at least don't say not to do this: https://mypy.readthedocs.io/en/latest/more_types.html#function-overloading

> I'm going to leave this plain.

Sure, sometimes it's not worth it.